File: instruction_decoding_8086/zig/disassembler_8086.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const File = std.fs.File;

const InstructionMode = enum(u2) {
    MemoryMode,
    MemoryMode8bitDisplacement,
    MemoryMode16bitDisplacement,
    RegisterMode,
};

const Registers = enum(u4) { al, ax, cl, cx, dl, dx, bl, bx, ah, sp, ch, bp, dh, si, bh, di };
const OpCodes = enum(u6) {
    mov = 34,
};

fn to_string(comptime T: type, value: T) []const u8 {
    const info = @typeInfo(T);
    if (info != .Enum) {
        @compileError("to_string: Expected an enum type");
    }
    const int_value = @intFromEnum(value);
    inline for (info.Enum.fields) |field| {
        if (int_value == @as(@TypeOf(int_value), field.value)) {
            return field.name;
        }
    }
    unreachable;
}

const Instruction = packed struct {
    r_m: u3,
    reg: u3,
    mod: InstructionMode,
    w: u1,
    d: u1,
    op_code: OpCodes,

    pub fn print(self: Instruction) void {
        std.debug.print("{b:0<8}_{b:0<8} => {b:0<6} {b:<1} {b:<1} {b:0<2} {b:0<3} {b:0<3}\n", .{
            @as(u16, @bitCast(self)) & @as(u16, 0xFF),
            @as(u16, @bitCast(self)) >> 8 & @as(u16, 0xFF),
            @intFromEnum(self.op_code),
            self.d,
            self.w,
            @intFromEnum(self.mod),
            self.reg,
            self.r_m,
        });
    }
};

fn read_data(allocator: Allocator, filename: []const u8) ![]u8 {
    const file = try std.fs.cwd().openFile(filename, File.OpenFlags{ .mode = .read_only });
    defer file.close();

    const file_stats = try file.stat();
    const buffer = try file.readToEndAlloc(allocator, file_stats.size);
    std.debug.assert(buffer.len == file_stats.size);
    return buffer;
}

fn parse_instruction(allocator: Allocator, data: []u8) ![]Instruction {
    if (std.mem.trim(u8, data, " \r\n\t").len & 1 != 0) {
        return error.InvalidInstructionSize;
    }
    const count = @as(f32, @floatFromInt(std.mem.trim(u8, data, " \r\n\t").len)) / 2.0;
    const instructions = try allocator.alloc(Instruction, @intFromFloat(count));
    for (instructions, 0..) |*instr, idx| {
        const raw_value = @as(u16, data[idx * 2]) << 8 | @as(u16, data[idx * 2 + 1]);

        instr.* = @bitCast(raw_value);

        // instr.* = Instruction{
        //     .op_code = @enumFromInt(@as(u6, @intCast((raw_value >> 10) & 0b111111))),
        //     .d = @intCast((raw_value >> 9) & 0b1),
        //     .w = @intCast((raw_value >> 8) & 0b1),
        //     .mod = @enumFromInt((raw_value >> 6) & 0b11),
        //     .reg = @intCast((raw_value >> 3) & 0b111),
        //     .r_m = @intCast(raw_value & 0b111),
        // };
    }
    return instructions;
}

fn disassemble(instructions: []Instruction) void {
    for (instructions) |instr| {
        var source: Registers = undefined;
        var destination: Registers = undefined;
        if (instr.d == 0) {
            destination = @enumFromInt(@as(u4, instr.r_m << 1) + instr.w);
            source = @enumFromInt(@as(u4, instr.reg << 1) + instr.w);
        } else {
            destination = @enumFromInt(@as(u4, instr.reg << 1) + instr.w);
            source = @enumFromInt(@as(u4, instr.r_m << 1) + instr.w);
        }
        std.debug.print("{s} {s}, {s}\n", .{
            to_string(OpCodes, instr.op_code),
            to_string(Registers, destination),
            to_string(Registers, source),
        });
    }
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);
    if (args.len == 2) {
        const filename = args[1];

        const data: []u8 = try read_data(allocator, filename);
        defer allocator.free(data);

        const instructions: []Instruction = try parse_instruction(allocator, data);
        defer allocator.free(instructions);

        for (instructions) |instruction| {
            instruction.print();
        }
        // disassemble(instructions);
    } else {
        std.debug.print("Usage: {s} <input_file>", .{args[0]});
    }
}

test "single_instruction" {
    const allocator = std.testing.allocator;
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);
    if (args.len == 2) {
        const filename = args[1];

        const data: []u8 = try read_data(allocator, filename);
        defer allocator.free(data);

        const instructions: []Instruction = try parse_instruction(allocator, data);
        defer allocator.free(instructions);

        disassemble(instructions);
    } else {
        std.debug.print("Usage: {s} <input_file>", .{args[0]});
    }
}

File: instruction_decoding_8086/c/disassembler_8086.c
#include <fcntl.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>

enum instruction_type {
    mov = 34,
};

union instruction {
    uint16_t value;
    struct __attribute__((packed)) {
        bool w : 1;
        bool d : 1;
        uint8_t op_code : 6;
        uint8_t r_m : 3;
        uint8_t reg : 3;
        uint8_t mod : 2;
    };
};

enum registers {
    al = 0,
    ax,
    cl,
    cx,
    dl,
    dx,
    bl,
    bx,
    ah,
    sp,
    ch,
    bp,
    dh,
    si,
    bh,
    di,
};

typedef union instruction instruction;
typedef enum registers registers;
typedef enum instruction_type instruction_type;

char *register_name(registers reg) {
    switch (reg) {
    case al:
        return "al";
    case ax:
        return "ax";
    case cl:
        return "cl";
    case cx:
        return "cx";
    case dl:
        return "dl";
    case dx:
        return "dx";
    case bl:
        return "bl";
    case bx:
        return "bx";
    case ah:
        return "ah";
    case sp:
        return "sp";
    case ch:
        return "ch";
    case bp:
        return "bp";
    case dh:
        return "dh";
    case si:
        return "si";
    case bh:
        return "bh";
    case di:
        return "di";
    }
}

char *instruction_name(instruction_type instr) {
    switch (instr) {
    case mov:
        return "mov";
    }
}

void disassemble(instruction instr) {
    registers destination;
    registers source;
    if (instr.d == 0) {
        destination = (instr.r_m << 1) + instr.w;
        source = (instr.reg << 1) + instr.w;
    } else {
        destination = (instr.reg << 1) + instr.w;
        source = (instr.r_m << 1) + instr.w;
    }
    printf("%s %s, %s\n", instruction_name((instruction_type)instr.op_code),
           register_name(destination), register_name(source));
}

int main(int argc, char **argv) {
    if (argc == 2) {
        char *filename = argv[1];
        int fd = open(filename, O_RDONLY);
        if (fd < 0) {
            perror("Failed to open file");
            return EXIT_FAILURE;
        }
        struct stat file_stats;
        if (fstat(fd, &file_stats) < 0) {
            perror("failed to get file size");
            return EXIT_FAILURE;
        }

        size_t file_size = file_stats.st_size;
        unsigned char *data =
            mmap(NULL, file_size, PROT_READ, MAP_PRIVATE, fd, 0);
        if (data == MAP_FAILED) {
            perror("mmap failed");
            close(fd);
            return EXIT_FAILURE;
        }
        int instruction_count = (float)file_size / sizeof(instruction);
        int idx = 0;
        instruction *instructions = (instruction *)data;
        while (idx < instruction_count) {
            disassemble(instructions[idx]);
            idx += 1;
        }
        close(fd);
    } else {
        printf("Usage: %s <input_file>\n", argv[0]);
    }
    return EXIT_SUCCESS;
}
